paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,518 @@
+  }
+
+  memory::desc CreateMemDescriptor(const std::vector<int> &dims,
+                                   memory::data_type type,
+                                   memory::format format) {
+    return platform::MKLDNNMemDesc(dims, type, format);
+  }
+
+  memory::desc CreateMemDescriptor(const Tensor *tensor, memory::data_type type,
+                                   memory::format format) {
+    auto dims = framework::vectorize2int(tensor->dims());
+    return CreateMemDescriptor(dims, type, format);
+  }
+
+  memory CreateMemory(const memory::desc &desc, const Tensor *tensor,
+                      bool is_y) {
+    if (is_y)
+      return CreateMemory(desc, tensor->data<K>());
+    else
+      return CreateMemory(desc, tensor->data<T>());
+  }
+
+  memory CreateMemory(const memory::desc &desc, const void *data) {
+    return memory({desc, engine_}, const_cast<void *>(data));
+  }
+
+  inner_product_forward CreateMulPrimitive(const memory &x_memory,
+                                           const memory &y_memory,
+                                           const memory::desc &dst_desc,
+                                           Tensor *output,
+                                           const ExecutionContext &ctx) {
+    const auto y_desc = y_memory.get_primitive_desc().desc();
+    const auto x_desc = x_memory.get_primitive_desc().desc();
+
+    auto mul_prim_desc = CreateMulPrimDesc(x_desc, y_desc, dst_desc);
+    output_ = this->CreateDstMemory(mul_prim_desc, ctx, output);
+
+    return inner_product_forward(mul_prim_desc, x_memory, y_memory, *output_);
+  }
+
+  inner_product_forward::primitive_desc CreateMulPrimDesc(
+      const memory::desc &x_desc, const memory::desc &y_desc,
+      const memory::desc &dst_desc) {
+    auto mul_desc = inner_product_forward::desc(prop_kind::forward, x_desc,
+                                                y_desc, dst_desc);
+
+    return inner_product_forward::primitive_desc(mul_desc, engine_);
+  }
+
+  memory CreateDstMemory(
+      const inner_product_forward::primitive_desc &mul_prim_desc,
+      const ExecutionContext &ctx, Tensor *output) {
+    auto dst_prim_desc = mul_prim_desc.dst_primitive_desc();
+    auto buffer_size = dst_prim_desc.get_size();
+
+    float *output_data =
+        output->mutable_data<float>(ctx.GetPlace(), buffer_size);
+    output->set_format((memory::format)dst_prim_desc.desc().data.format);
+    return memory(dst_prim_desc, to_void_cast<float>(output_data));
+  }
+
+ protected:
+  const mkldnn::engine &engine_;
+  boost::optional<memory> x_input_;
+  boost::optional<memory> y_input_;
+  boost::optional<memory> output_;
+  boost::optional<inner_product_forward> mul_;
+};  // namespace operators
+
+template <typename T, typename K>
+class QuantMulPrimitiveFactory : public MulPrimitiveFactory<T, K> {
+ public:
+  using MulPrimitiveFactory<T, K>::MulPrimitiveFactory;
+
+  virtual inner_product_forward CreateMulPrimitive(
```

Comment:
This function has gotten quite lengthy now. You may consider dividing it into functions such as:
`CreateInputMemories()`, `CreateXMemory()`, `CreateYMemory()`, `CreateOutputMemory()` so that their outputs can be fed into `CreateMulPrimitive()`. Nothing crucial, but it would improve the readability of the code.

Reply:
I have refined the code to make this function shorter.